[BEAM-52] Kafka custom source

rangadi opened the pull request:
Current status (3/18/16) : well tested and feature complete. going through reviews.

--
This is a preliminary PR. It is not tested yet. I am working on example application and unit tests. Some TODOs : 

- [x] unit tests
- [x] testing in the cluster with larger volumes of data and larger number of partitions
- [x] add more stats (I need to look into stats support in sdk)

rangadi commented:
I signed the CLA.

rangadi commented:
The functionality is well tested. will update javadoc today.

dhalperi commented on contrib/kafka/src/main/java/com/google/cloud/dataflow/contrib/kafka/KafkaCheckpointMark.java:
```diff
@@ -0,0 +1,64 @@
+ * Checkpoint for an unbounded KafkaSource reader. Consists of Kafka topic name, partition id,
+ * and the latest offset consumed so far.
+ *
+ * @author rangadi
```
We tend not to use these.

rangadi replied:
I like the policy. removed @author tags.

rangadi commented:
updates:

- move TopHashtagsExample to separate directory hierarchy
- many checkstyle and javadoc fixes 
- invoke watermarkFn with null record if the reader hasn't read any records.

dhalperi commented on contrib/kafka/src/main/java/com/google/cloud/dataflow/contrib/kafka/KafkaSource.java:
```diff
@@ -0,0 +1,806 @@
+/*
```
Could you rename this file `KafkaIO` and have the `PTransform<PBegin, [output type]>` be `KafkaIO.Read` or something similar? This follows examples like in `BigQueryIO`, `PubSubIO`, `BigtableIO`.

Hiding the source from the caller and exposing PTransforms generally makes the sources more usable and also allows for evolving the implementation over time.

(As an example, I'd like to rewrite `DatastoreIO.Read` based on `Create` instead of `BoundedSource`, but since we exposed the `BoundedSource` it's hard to do in a backwards-compatible way.)

dhalperi commented on contrib/kafka/src/main/java/com/google/cloud/dataflow/contrib/kafka/KafkaSource.java:
```diff
@@ -0,0 +1,806 @@
+ *
+ * <pre>
+ * Usage:
+ *        UnboundedSource&lt;String, ?&gt; kafkaSource = KafkaSource
```
I had thought that in our initial discussion the plan was to implement both a Bounded and Unbounded source version.

For now, using `UnboundedSource` in conjunction with `BoundedReadFromUnbounded` source makes sense, but -- if it makes sense -- it will eventually be nicer to do both so that the BoundedSource can also use dynamic work rebalancing.

dhalperi replied:
(and again, exposing a PTransform instead of an UnboundedSource would seem to make this easier -- the code can switch between the two implentations in `.apply`)

rangadi replied:
I agree, a bounded source would be quite useful. 
will also look into making it a PTransform.

dhalperi commented on contrib/kafka/src/main/java/com/google/cloud/dataflow/contrib/kafka/KafkaSource.java:
```diff
@@ -0,0 +1,806 @@
+
+  private static final Logger LOG = LoggerFactory.getLogger(KafkaSource.class);
+
+  private static class NowTimestampFn<T> implements SerializableFunction<T, Instant> {
```
This really ought to be in a TimestampFunctions library,

dhalperi replied:
(that note is more for me than for you.. sorry)

rangadi replied:
these don't look as bad with Java 8.

dhalperi commented on contrib/kafka/src/main/java/com/google/cloud/dataflow/contrib/kafka/KafkaSource.java:
```diff
@@ -0,0 +1,806 @@
+
+  /**
+   * Similar to {@link #unboundedSourceBuilder()}, where user in only interested in value, and
+   * wants to discard Kafak record key and metadata.
```
Kafka

dhalperi commented on contrib/kafka/src/main/java/com/google/cloud/dataflow/contrib/kafka/KafkaSource.java:
```diff
@@ -0,0 +1,806 @@
+ *            .withValueCoder(StringUtf8Coder.of())
+ *            .withTimestampFn(timestampFn)
+ *            .withWatermarkFn(watermarkFn)
+ *            .build();
```
It seems like there are three different levels of abstraction the user may want from their Kafka reader:

1. `KV<K, V>`  (keys and values)
2. `V`  (values)
3. `KafkaRecord<K, V>`  (keys and values, plus metadata about the record).

based on the Kafka docs I read [here](http://kafka.apache.org/documentation.html#connect_streamsandrecords) I would probably suggest that 1 is the default, and there should be a `Read#withMetadata` builder that turns it into 3.

I'm not sure that 2 is necessary at all -- it's easy to turn a `PCollection<KV<K,V>>` into a `PCollection<V>` using the [`Values` transform](https://github.com/GoogleCloudPlatform/DataflowJavaSDK/blob/master/sdk/src/main/java/com/google/cloud/dataflow/sdk/transforms/Values.java#L58)

Thoughts?

dhalperi commented on contrib/kafka/src/main/java/com/google/cloud/dataflow/contrib/kafka/KafkaSource.java:
```diff
@@ -0,0 +1,806 @@
+    }
+  }
+
+  public static <K, V> Builder<K, V> unboundedSourceBuilder() {
```
would think this is just `KafkaIO.read()` static factory method for `KafkaIO.Read` class.

dhalperi commented on contrib/kafka/src/main/java/com/google/cloud/dataflow/contrib/kafka/KafkaSource.java:
```diff
@@ -0,0 +1,806 @@
+    return new Builder<K, V>();
+  }
+
+  public static Builder<byte[], byte[]> unboundedByteSourceBuilder() {
```
Maybe ought to be the default values for `K`, `V` in `KafkaIO.read()`, right?

dhalperi replied:
(or string, and then this can still be constructed using `.withKeyCoder`/`.withValueCoder`

dhalperi commented on contrib/kafka/src/main/java/com/google/cloud/dataflow/contrib/kafka/KafkaSource.java:
```diff
@@ -0,0 +1,806 @@
+   * @param K key type
+   * @param V value type
+   */
+  public static class Builder<K, V> {
```
This is totally ad-libbed, but what about something like this:

```java
public class KafkaIO {

  public static class Read< K, V> extends PTransform<PBegin, PCollection<KV<K, V>>> {
    private Read(Coder<K> keyCoder, Coder<V> valueCoder, ...) {
    }

    public ReadWithMetadata<K, V> withMetadata() {
        return new ReadWithMetadata(...);
    }
  }

  public static class ReadWithMetadata< K, V> extends PTransform<PBegin, PCollection<KafkaRecord<K, V>>> {
    private ReadWithMetadata(Coder<K> keyCoder, Coder<V> valueCoder, ...) {...}

    public Read<K, V> withoutMetadata() {
        return new Read(...);
    }
  }
}
```

rangadi replied:
will look into this.

dhalperi commented on contrib/kafka/src/main/java/com/google/cloud/dataflow/contrib/kafka/KafkaSource.java:
```diff
@@ -0,0 +1,806 @@
+     * @return Builder
+     */
+    public Builder<K, V> withConsumerProperty(String configKey, Object configValue) {
+      Preconditions.checkArgument(!ignoredConsumerProperties.containsKey(configKey),
```
Following [the advice of the Guava team](https://github.com/google/guava/wiki/PreconditionsExplained), we usually import these statically.

```java

import static ...Preconditions.checkArgument;
...
checkArgument(!ignoredConsumerProperties.containsKey(configKey), "message");
```

dhalperi commented on contrib/kafka/src/main/java/com/google/cloud/dataflow/contrib/kafka/KafkaSource.java:
```diff
@@ -0,0 +1,806 @@
+      Preconditions.checkArgument(!ignoredConsumerProperties.containsKey(configKey),
+          "No need to configure '%s'. %s", configKey, ignoredConsumerProperties.get(configKey));
+      mutableConsumerConfig.put(configKey, configValue);
+      return this;
```
we strongly discourage mutating these builders. Instead, it probably makes sense to just take a fully-formed `Map<String, Object>`.

dhalperi commented on contrib/kafka/src/main/java/com/google/cloud/dataflow/contrib/kafka/KafkaSource.java:
```diff
@@ -0,0 +1,806 @@
+
+    /**
+     * Update consumer config properties. Note that this does not not discard already configured.
+     * Same as invoking #withConsumerProperty() with each entry.
```
`{@link #withConsumerProperty}` or something like that ?

dhalperi commented on contrib/kafka/src/main/java/com/google/cloud/dataflow/contrib/kafka/KafkaSource.java:
```diff
@@ -0,0 +1,806 @@
+    }
+
+    /**
+     * Set Coder for Key.
```
`{@link Coder}`

dhalperi commented on contrib/kafka/src/main/java/com/google/cloud/dataflow/contrib/kafka/KafkaSource.java:
```diff
@@ -0,0 +1,806 @@
+     * Update consumer config properties. Note that this does not not discard already configured.
+     * Same as invoking #withConsumerProperty() with each entry.
+     *
+     * @param configUpdate updates to {@link ConsumerConfig}
```
generally, feel free to drop `@param` and `@return` tags -- relatively unused in google style.

rangadi replied:
I didn't have those initially. ```mvn javadoc:javadoc``` warns without those.

dhalperi commented on contrib/kafka/src/main/java/com/google/cloud/dataflow/contrib/kafka/KafkaSource.java:
```diff
@@ -0,0 +1,806 @@
+
+    /**
+     * A function to assign a timestamp to a record. When this is not set, processing timestamp
+     * (when record is processed by {@link UnboundedReader#advance()}) is used.
```
better just "defaults to processing time".

dhalperi commented on contrib/kafka/src/main/java/com/google/cloud/dataflow/contrib/kafka/KafkaSource.java:
```diff
@@ -0,0 +1,806 @@
+     */
+    public UnboundedSource<KafkaRecord<K, V>, KafkaCheckpointMark> build() {
+
+      ImmutableMap<String, Object> consumerConfig = ImmutableMap.copyOf(mutableConsumerConfig);
```
I don't believe this is deeply immutable, so may not actually be immutable.

dhalperi commented on contrib/kafka/src/main/java/com/google/cloud/dataflow/contrib/kafka/KafkaSource.java:
```diff
@@ -0,0 +1,806 @@
+    }
+
+    @Override
+    public List<UnboundedKafkaSource<K, V>> generateInitialSplits(
```
probably worth javadoccing the splitting behavior extensively here, and briefly at the class level (with a `{@link }` pointer to here). Users will want to understand how the source is scalabale

rangadi replied:
yep. that is important. I will add.

dhalperi commented on contrib/kafka/src/main/java/com/google/cloud/dataflow/contrib/kafka/KafkaSource.java:
```diff
@@ -0,0 +1,806 @@
+        }
+      });
+
+      Preconditions.checkArgument(desiredNumSplits > 0);
```
message

dhalperi commented on contrib/kafka/src/main/java/com/google/cloud/dataflow/contrib/kafka/KafkaSource.java:
```diff
@@ -0,0 +1,806 @@
+
+      int numSplits = Math.min(desiredNumSplits, partitions.size());
+
+      List<List<TopicPartition>> assignments = Lists.newArrayList();
```
just `new ArrayList<>();` -- the Guava team's recommendation is to stop using these static factory methods since Java 7 has the diamond (`<>`) syntax.

rangadi replied:
I see. will remove them. Diamond systax does not seem to be effect except for defintions. E.g. ```x.addList(new ArrayList<>())``` requires explicit type

dhalperi commented on contrib/kafka/src/main/java/com/google/cloud/dataflow/contrib/kafka/KafkaSource.java:
```diff
@@ -0,0 +1,806 @@
+    }
+
+    private static <T> T decode(byte[] bytes, Coder<T> coder) throws IOException {
+      if (bytes == null) {
```
can bytes be null? That seems strange. What would that mean?

rangadi replied:
'key' could be null. not sure about value (mostly not).

dhalperi commented on contrib/kafka/src/main/java/com/google/cloud/dataflow/contrib/kafka/KafkaSource.java:
```diff
@@ -0,0 +1,806 @@
+   * Usually the users are only interested in value in KafkaRecord. This is a convenient class
+   * to strip out other fields in KafkaRecord returned by UnboundedKafkaValueSource
+   */
+  private static class UnboundedKafkaValueSource<V>
```
Oh man. There has to be a way to make it possible that we only have a single implementation of the Source/Reader but we can support both with/without record metadata. I'll think and get back to you.

rangadi replied:
Yeah. Since user provides multiple functions, the type affects the builder/reader etc.. I didn't see a good way to avoid these wrappers.

rangadi commented:
added unit tests (CountingSourceTest adapted to KafkaSource).

rangadi commented:
multiple updates
- KafkaIO refactor : interface changes to return a transform (biggest part of change).
- allow specifying specific partitions (rather than just topics)
- use KV in KafkaRecord
- address most or all of review comments from Dan.

dhalperi commented:
@rangadi I know you've made some threading and performance improvements in the last day or so. Ping me here once the code is ready again.

dhalperi commented on contrib/examples/kafka/src/main/java/com/google/cloud/dataflow/contrib/kafka/examples/TopHashtagsExample.java:
```diff
@@ -0,0 +1,262 @@
+    @Description("Sliding window size, in minutes")
+    @Default.Integer(10)
+    Integer getSlidingWindowSize();
+    void setSlidingWindowSize(Integer value);
```
@kennknowles any suggestions on naming here? Maybe `--slidingWindowLengthMinutes` and `--slidingWindowIntervalMinutes`?

@rangadi in general, prefer units in variables when they are not super obvious. E.g., `timeoutMs` instead of just `timeout`

dhalperi commented on contrib/examples/kafka/src/main/java/com/google/cloud/dataflow/contrib/kafka/examples/TopHashtagsExample.java:
```diff
@@ -0,0 +1,262 @@
+    Integer getSlidingWindowPeriod();
+    void setSlidingWindowPeriod(Integer value);
+
+    @Description("Bootstarp Server(s) for Kafka")
```
typo bootstrap

dhalperi commented on contrib/examples/kafka/src/main/java/com/google/cloud/dataflow/contrib/kafka/examples/TopHashtagsExample.java:
```diff
@@ -0,0 +1,262 @@
+    String getBootstrapServers();
+    void setBootstrapServers(String servers);
+
+    @Description("One or more topics to read from")
```
".. comma separated"

dhalperi commented on contrib/examples/kafka/src/main/java/com/google/cloud/dataflow/contrib/kafka/examples/TopHashtagsExample.java:
```diff
@@ -0,0 +1,262 @@
+    final int windowSize = options.getSlidingWindowSize();
+    final int windowPeriod = options.getSlidingWindowPeriod();
+
+    Read<?, String> reader = KafkaIO.read()
```
prefer `KafkaIO.Read` rather than static importing `Read`. Lots of things have a `Read` class so it makes the snippet clearer.

rangadi commented:
Updated the patch. 'Length Minutes' and 'Interval Minutes' sound better than 'Size' and 'Period' for window arguments.

dhalperi commented on contrib/examples/kafka/src/main/java/com/google/cloud/dataflow/contrib/kafka/examples/TopHashtagsExample.java:
```diff
@@ -0,0 +1,257 @@
+    String getBootstrapServers();
+    void setBootstrapServers(String servers);
+
+    @Description("One or more comma seperated topics to read from")
```
typo: separated.

dhalperi commented on contrib/examples/kafka/src/main/java/com/google/cloud/dataflow/contrib/kafka/examples/TopHashtagsExample.java:
```diff
@@ -0,0 +1,257 @@
+        .withTimestampFn(timestampFn);
+
+    pipeline
+      .apply("sample_tweets", reader)
```
why only supply a name for this stage?

rangadi replied:
I had used this name with previous `Read.from(...).named()`. We could remove it.

dhalperi commented on contrib/examples/kafka/src/main/java/com/google/cloud/dataflow/contrib/kafka/examples/TopHashtagsExample.java:
```diff
@@ -0,0 +1,257 @@
+
+  // The rest of the file implements DoFns to extract hashtags, formatting output, writing output
+  // back to Kafka. Note that writing to Kafka is not a complete Dataflow Sink. It is a best-effort
+  // logging of the results.
```
Take a grammar pass over this paragraph?

Drop the bit about the sink? That doc goes on KafkaIO.Write if anywhere.

rangadi replied:
rephrased and removed reference to sink.

dhalperi commented on contrib/examples/kafka/src/main/java/com/google/cloud/dataflow/contrib/kafka/examples/TopHashtagsExample.java:
```diff
@@ -0,0 +1,257 @@
+      for (JsonNode hashtag : jsonMapper.readTree(ctx.element())
+          .with("entities")
+          .withArray("hashtags")) {
+        ctx.output(hashtag.get("text").asText());
```
You might want to check your Eclipse configuration. It is making some wonky choices with respect to indentation etc. Obviously we're not going to enforce Google style in Beam, but we'd welcome it for contributions from Googlers like yourself ;)

Here's an open-source copy: https://github.com/google/styleguide/blob/gh-pages/eclipse-java-google-style.xml

dhalperi commented on contrib/examples/kafka/src/main/java/com/google/cloud/dataflow/contrib/kafka/examples/TopHashtagsExample.java:
```diff
@@ -0,0 +1,257 @@
+    public void processElement(ProcessContext ctx) throws Exception {
+      for (JsonNode hashtag : jsonMapper.readTree(ctx.element())
+          .with("entities")
+          .withArray("hashtags")) {
```
does Jackson gracefully handle the case when `entities` or `entities.hashtags` is missing, or will it crash? Do you want to catch and ignore?

dhalperi commented on contrib/examples/kafka/src/main/java/com/google/cloud/dataflow/contrib/kafka/examples/TopHashtagsExample.java:
```diff
@@ -0,0 +1,257 @@
+  }
+
+  // extract timestamp from "timestamp_ms" field.
+  private static SerializableFunction<KV<byte[], String>, Instant> timestampFn =
```
`final`?

Maybe name it something like `TWEET_TIMESTAMP_OR_NOW`?

Maybe Javadoc it to explain what it does (timestamp or now)?

rangadi replied:
done.

dhalperi commented on contrib/examples/kafka/src/main/java/com/google/cloud/dataflow/contrib/kafka/examples/TopHashtagsExample.java:
```diff
@@ -0,0 +1,257 @@
+  private static class OutputFormatter extends DoFn<List<KV<String, Long>>, String>
+      implements DoFn.RequiresWindowAccess {
+
+    private transient DateTimeFormatter formatter;
```
why not make this `private static final` rather than re-allocating per bundle? `DateTimeFormatter` is threadsafe, right?

rangadi replied:
done. that is better.

dhalperi commented on contrib/examples/kafka/src/main/java/com/google/cloud/dataflow/contrib/kafka/examples/TopHashtagsExample.java:
```diff
@@ -0,0 +1,257 @@
+      implements DoFn.RequiresWindowAccess {
+
+    private transient DateTimeFormatter formatter;
+    private transient ObjectWriter jsonWriter;
```
same for `jsonWriter` -- also definitely threadsafe.

dhalperi commented on contrib/examples/kafka/src/main/java/com/google/cloud/dataflow/contrib/kafka/examples/TopHashtagsExample.java:
```diff
@@ -0,0 +1,257 @@
+    @Override
+    public void processElement(ProcessContext ctx) throws Exception {
+      if (producer == null) {
+        producer = new KafkaProducer<String, String>(config);
```
Just allocate `producer` in `startBundle`, and keep the code simple.

rangadi replied:
done. and should I close the producer in finishBundle()?

dhalperi commented on contrib/examples/kafka/src/main/java/com/google/cloud/dataflow/contrib/kafka/examples/TopHashtagsExample.java:
```diff
@@ -0,0 +1,257 @@
+      if (producer == null) {
+        producer = new KafkaProducer<String, String>(config);
+      }
+      LOG.info("Top Hashtags : " + ctx.element());
```
Do not log per element -- super slow!

Just drop I think.

(If you were going to log, it would be `LOG.info("Top hashtags: {}", ctx.element());` -- for performance when logging is disabled.

dhalperi commented on contrib/kafka/pom.xml:
```diff
@@ -0,0 +1,176 @@
+    <google-cloud-dataflow-version>[1.2.0,2.0.0)</google-cloud-dataflow-version>
+    <hamcrest.version>1.3</hamcrest.version>
+    <junit.version>4.11</junit.version>
+    <slf4j.version>1.7.7</slf4j.version>
```
are these not inherited from the parent?

I'd imagine I'll drop all these when I import to Beam

rangadi replied:
This is not addressed yet. I had some issue trying to use parent pom. I will check.

dhalperi commented on contrib/kafka/src/main/java/com/google/cloud/dataflow/contrib/kafka/KafkaCheckpointMark.java:
```diff
@@ -0,0 +1,80 @@
+     * no-duplicates, it could support Dataflow restart better. Unlike an update of a dataflow job,
+     * a restart does not have checkpoint state. This secondary checkpoint might be a good start
+     * for readers. Another similar benefit is when the number of workers or number of
+     * Kafka partitions changes.
```
I would, via syntax, logically separate these two points.

```java
... {
    /* do nothing */

    // TODO: We might want to support committing the consumer offset to the Kafka broker. This would mean that
    // if we restarted the job fresh we would not get the old data.
```

I honestly did not actually understand your paragraph, so the above is my best guess. A few questions about what you wrote:

* would the restarted job use the same consumer groups?
* why not just let the user specific an optional offset for each topic/partition instead? That functionality seems like it would completely subsume any benefits of committing the consumer offset. Not that I think you should make this change -- I don't -- but I'm trying to understand the rationale for this TODO vs other possibilities.

rangadi replied:
That is correct. I rewrote the comment close to what you suggested. 
 - The restarted job need to use the same consumer group ("client_id")
 - yes, this is effectively same as letting users specifying an offset for a partition. 

The third case is handling the case when the number of partitions or Dataflow splits changes across an update. The committed offsets will let us handle changes in distribution of partitions among splits. Thats what the latter part of the my comment was about.

dhalperi commented on contrib/kafka/src/main/java/com/google/cloud/dataflow/contrib/kafka/KafkaCheckpointMark.java:
```diff
@@ -0,0 +1,80 @@
+
+  /**
+   * TopicPartition, offset tuple. Defines specific location in the partitions.
+   */
```
doesn't parse -- try with complete sentences?

I think this is just a bean-like object to hold the topic, partition, and offset that comprise the checkpoint mark for a single partition. Easy to write out as a whole sentence. :)

rangadi replied:
Yep, that is better.

dhalperi commented on contrib/kafka/src/main/java/com/google/cloud/dataflow/contrib/kafka/KafkaCheckpointMark.java:
```diff
@@ -0,0 +1,80 @@
+import java.util.List;
+
+/**
+ * Checkpoint for an unbounded KafkaSource reader. Consists of Kafka topic name, partition id,
```
Checkpoint for an unbounded `KafkaIO.Read`.

dhalperi commented on contrib/kafka/src/main/java/com/google/cloud/dataflow/contrib/kafka/KafkaIO.java:
```diff
@@ -0,0 +1,928 @@
+
+import javax.annotation.Nullable;
+
+/**
```
please flesh out. Use https://github.com/apache/incubator-beam/blob/master/sdk/src/main/java/com/google/cloud/dataflow/sdk/io/bigtable/BigtableIO.java#L66 as a reference.

rangadi replied:
Yes, BigtableIO JavaDoc is so much better. I need spend more time on this. will get back to it.

rangadi replied:
Updated JavaDoc with code sample with examples.

dhalperi commented on contrib/kafka/src/main/java/com/google/cloud/dataflow/contrib/kafka/KafkaIO.java:
```diff
@@ -0,0 +1,928 @@
+ * {@link UnboundedKafkaSource#generateInitialSplits(int, PipelineOptions)} for more details on
+ * splits and checkpoint support.
+ *
+ * <pre>{@code
```
```
<h3>Reading from Kafka</h3>

To read from a Kafka broker, you must configure the bootstrap servers and one or more Kafka topics to consume.

<pre>{@code
Pipeline p = ...;
PCollection<byte[], byte[]> unboundedMessages =
    p.apply("Read two topic_a and topic_b",
        KafkaIO.read().withBootstrapServers("..").withTopics(...));
...
}</pre>

<p>You may optionally supply a `Coder<V>` value coder to determine the decoding of the message:

<pre>{@code
PCollection<byte[], V> messages = p.apply...
}</pre>
...

<p>To produce a bounded {@code PCollection}, use {@link KafkaIO.Read#withMaxNumRecords} or
...:

<pre>{@code
...
}</pre>

```

etc.

A few guidelines:

* Spell out the basic use cases in logical separation (source before sink, unbounded before bounded) and in ascending order of complexity (default configuration, overriding the value coder, overriding the key coder, supplying custom timestamps). Do not do it all at once.

* Try to put the minimal content that is both necessary (include that `p` is a `Pipeline`) and sufficient (drop the `Values.<String>create()`).

Any gotchas you can think of, document them here.

rangadi replied:
Thanks! this will be very useful. will do.

dhalperi commented on contrib/kafka/src/main/java/com/google/cloud/dataflow/contrib/kafka/KafkaIO.java:
```diff
@@ -0,0 +1,928 @@
+        Coder<K> keyCoder,
+        Coder<V> valueCoder,
+        @Nullable SerializableFunction<KV<K, V>, Instant> timestampFn,
+        @Nullable SerializableFunction<KV<K, V>, Instant> watermarkFn,
```
I would not make these `null` -- I would instead put the defaults in the `KafkaIO.read()` static factory method.

I would in fact enforce that they are not `null`.

dhalperi replied:
note that I now reject this, given the issues with changing different typed parameters at the same time.

dhalperi commented on contrib/kafka/src/main/java/com/google/cloud/dataflow/contrib/kafka/KafkaIO.java:
```diff
@@ -0,0 +1,928 @@
+     * A factory to create Kafka {@link Consumer} from consumer configuration.
+     * Mainly used for tests. Default is {@link KafkaConsumer}.
+     */
+    public Read<K, V> withConsumerFactoryFn(
```
if it's only used for tests, make it package-private and annotate with `@VisibleForTesting`.

rangadi replied:
This is required to support another version of the consumer (e.g. Kafka 0.8.x Consumer).

dhalperi commented on contrib/kafka/src/main/java/com/google/cloud/dataflow/contrib/kafka/KafkaIO.java:
```diff
@@ -0,0 +1,928 @@
+
+    // default Kafka 0.9 Consumer supplier. static variable to avoid capturing 'this'
+    private static SerializableFunction<Map<String, Object>, Consumer<byte[], byte[]>>
+      kafka9ConsumerFactory =
```
Why isn't this `private static final KAFKA_9_CONSUMER_FACTORY_FN`?

dhalperi replied:
https://google.github.io/styleguide/javaguide.html#s5.2.4-constant-names

dhalperi commented on contrib/kafka/src/main/java/com/google/cloud/dataflow/contrib/kafka/KafkaIO.java:
```diff
@@ -0,0 +1,928 @@
+      SerializableFunction<Map<String, Object>, Consumer<byte[], byte[]>> consumerFactoryFn;
+    private final Map<String, Object> consumerConfig;
+    private final long maxNumRecords; // bounded read, mainly for testing
+    private final Duration maxReadTime; // bounded read, mainly for testing
```
drop the "mainly for testing" part -- it may be true as far as we know, but we probably don't want to put value judgements on the user's pipeline decisions.

(Btw, I acknowledge that I think I've written comments that violate what I just wrote in the past. With time comes wisdom.... or so I hope)

dhalperi commented on contrib/kafka/src/main/java/com/google/cloud/dataflow/contrib/kafka/KafkaIO.java:
```diff
@@ -0,0 +1,928 @@
+        );
+
+    // set config defaults
+    private static final Map<String, Object> defaultConsumerProperties =
```
DEFAULT_CONSUMER_PROPERTIES

dhalperi commented on contrib/kafka/src/main/java/com/google/cloud/dataflow/contrib/kafka/KafkaIO.java:
```diff
@@ -0,0 +1,928 @@
+    /**
+     * A set of properties that are not required or don't make sense for our consumer.
+     */
+    private static final Map<String, String> ignoredConsumerProperties = ImmutableMap.of(
```
IGNORED_CONSUMER_PROPERTIES

dhalperi replied:
needs work

dhalperi commented on contrib/kafka/src/main/java/com/google/cloud/dataflow/contrib/kafka/KafkaIO.java:
```diff
@@ -0,0 +1,928 @@
+     */
+    public <KeyT> Read<KeyT, V> withKeyCoder(Coder<KeyT> keyCoder) {
+      checkState(timestampFn == null, "Set timestampFn after setting key and value coders");
+      checkState(watermarkFn == null, "Set watermarkFn after setting key and value coders");
```
okay, I see this may invalidate some earlier comments of mine. this is definitely an interesting and thorny issue...

let's sync tomororw, but you may well have found a perfectly reasonable solution to this :)

dhalperi commented on contrib/kafka/src/main/java/com/google/cloud/dataflow/contrib/kafka/KafkaIO.java:
```diff
@@ -0,0 +1,928 @@
+  }
+
+  /**
+   * Creates and uninitialized {@link Read} {@link PTransform}. Before use, basic Kafka
```
an

rangadi commented:
Wow! thanks for such a detailed review. I will address it.

Added support for reporting backlog. Since I added one thread for kafka I/O anyway, added one more for fetching offsets.

rangadi commented:
@dhalperi Updated Read interface as we discussed. Needed changes to tests as well.
TODO: good JavaDoc for KafkaIO as you suggested. will update later today.

dhalperi commented on sdk/src/main/java/com/google/cloud/dataflow/sdk/io/bigtable/BigtableIO.java:
```diff
@@ -72,7 +72,7 @@
  * <h3>Reading from Cloud Bigtable</h3>
  *
  * <p>The Bigtable source returns a set of rows from a single table, returning a
- * {@code PCollection&lt;Row&gt;}.
```
please revert

rangadi replied:
done. Eclipse didn't render this properly.

dhalperi commented on contrib/examples/kafka/src/main/java/com/google/cloud/dataflow/contrib/kafka/examples/TopHashtagsExample.java:
```diff
@@ -0,0 +1,265 @@
+
+    @Override
+    public void finishBundle(Context c) throws Exception {
+      producer.close();
```
If the producer is closed in `finishBundle` will it need to be reopened in `startBundle`? Is there a sort of `flush()` instead?

rangadi replied:
There is a `flush()`. But we would still end up with many unclosed producers. How about making it static?

dhalperi commented on contrib/examples/kafka/src/main/java/com/google/cloud/dataflow/contrib/kafka/examples/TopHashtagsExample.java:
```diff
@@ -0,0 +1,265 @@
+
+    @Override
+    public void processElement(ProcessContext ctx) throws Exception {
+      LOG.info("Top Hashtags : {}", ctx.element());
```
Logging per element is considered harmful! Remove or downgrade to trace?

rangadi replied:
done. I logged it since it is only once every minute by default.

dhalperi commented on contrib/kafka/src/main/java/com/google/cloud/dataflow/contrib/kafka/KafkaIO.java:
```diff
@@ -0,0 +1,1046 @@
+    public UnboundedKafkaReader<K, V> createReader(PipelineOptions options,
+                                                   KafkaCheckpointMark checkpointMark) {
+      if (assignedPartitions.isEmpty()) {
+        LOG.warn("hack: working around DirectRunner issue. It does not generateSplits()");
```
@tgroh opinions here?

tgroh replied:
This message shouldn't be at `WARN`; potentially keep, but change the message to note that `generateInitialSplits` wasn't called, which will cause (**BAD THINGS**) to occur, or move to a lower log level

As there's no requirement to actually generate splits, this is instead a fallback to initialize in the case where we call the initial source. Given that it appears that you need the Consumer, which may require information only available at execution time, this seems like a reasonable workaround for that - if it's possible to generate some shorthand `TopicPartition` for each topic that is "100% of this topic", it would be preferable to construct that at source construction time and put it in the `assignedPartitions`, and then remove this block

In the future, we plan on exercising `generateInitialSplits` in the `InProcessPipelineRunner`, but may skip it if the number of target splits is equal to 1, so this would likely have to stick around to handle that case.

rangadi replied:
`generateInitialSplits()` is public API and it is better if the contract is clear about when it is called. [ My personal preference is not to skip it in DirectRunning  (it can run multiple splits in sequence if more than one is returned), but that is a separate issue, what if I have a source that does not want to have less than 3 splits?]

dhalperi replied:
A little more context:

This is not just the `DirectPipelineRunner` -- any runner may choose not to call `generateInitialSplits`, and the source needs to support reading in that case.

I might follow Thomas' suggestion and go with something like

```java
LOG.warn("It looks like the runner did not call generateInitialSplits on this {}. This means that the Kafka stream will not be consumed in parallel.", this);
```

rangadi replied:
Yep. I was going to update the message.
still I am not a fan of selectively not calling `generateInitialSplits()`. I think is better for the runner to figure out how to handle multiple splits than requiring source developers to special case it. Related issue : a source can return more splits in `desiredSplits`.

dhalperi replied:
needs work.

dhalperi commented on contrib/kafka/src/main/java/com/google/cloud/dataflow/contrib/kafka/KafkaIO.java:
```diff
@@ -0,0 +1,1046 @@
+
+    @Override
+    public boolean requiresDeduping() {
+      return false;
```
should comment why this is false.

rangadi replied:
done

dhalperi commented on contrib/kafka/src/main/java/com/google/cloud/dataflow/contrib/kafka/KafkaIO.java:
```diff
@@ -0,0 +1,1046 @@
+          TopicPartition assigned = source.assignedPartitions.get(i);
+
+          checkState(ckptMark.getTopicPartition().equals(assigned),
+              "checkpointed partition %s and assinged partition %s don't match at position %d",
```
typo: assigned

dhalperi replied:
needs work

dhalperi commented on contrib/kafka/src/main/java/com/google/cloud/dataflow/contrib/kafka/KafkaIO.java:
```diff
@@ -0,0 +1,1046 @@
+      // Read in a loop and enqueue the batch of records, if any, to availableRecordsQueue
+      while (!closed) {
+        try {
+          ConsumerRecords<byte[], byte[]> records = consumer.poll(1000);
```
extract 1000 as constant and comment

dhalperi commented on contrib/kafka/src/main/java/com/google/cloud/dataflow/contrib/kafka/KafkaIO.java:
```diff
@@ -0,0 +1,1046 @@
+            availableRecordsQueue.put(records); // blocks until dequeued.
+          }
+        } catch (InterruptedException e) {
+          LOG.warn(this + " consumer thread is interrupted", e); // not expected
```
use `LOG.warn("{} consumer thread", this, e);` instead

dhalperi replied:
or better yet make all the logs in this file have the same scheme. Below it's `Log.WARN("{}: ...");`

rangadi replied:
Done. Updated the all the LOG lines in the reader. 
JavaDoc was not clear about how exception is rendered with multiple objects as in `LOG.warn("msg", o1, exception)`. Just tried it, it does the right thing.

dhalperi replied:
http://stackoverflow.com/a/6374166/1715495

On Tue, Apr 5, 2016 at 10:38 AM, Raghu Angadi <notifications@github.com>
wrote:

> In
> contrib/kafka/src/main/java/com/google/cloud/dataflow/contrib/kafka/KafkaIO.java
> <https://github.com/GoogleCloudPlatform/DataflowJavaSDK/pull/121#discussion_r58581720>
> :
>
> > +
> > +          partitionStates.get(i).consumedOffset = ckptMark.getOffset();
> > +        }
> > +      }
> > +    }
> > +
> > +    private void consumerPollLoop() {
> > +      // Read in a loop and enqueue the batch of records, if any, to availableRecordsQueue
> > +      while (!closed) {
> > +        try {
> > +          ConsumerRecords<byte[], byte[]> records = consumer.poll(1000);
> > +          if (!records.isEmpty()) {
> > +            availableRecordsQueue.put(records); // blocks until dequeued.
> > +          }
> > +        } catch (InterruptedException e) {
> > +          LOG.warn(this + " consumer thread is interrupted", e); // not expected
>
> Done. Updated the all the LOG lines in the reader.
> JavaDoc was not clear about how exception is rendered with multiple
> objects as in LOG.warn("msg", o1, exception). Just tried it, it does the
> right thing.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly or view it on GitHub
> <https://github.com/GoogleCloudPlatform/DataflowJavaSDK/pull/121/files/3d560732adfb56d572740985bd4bc3df7900b1e3#r58581720>
>

rangadi replied:
Thanks. Confirmed the behavior.

dhalperi commented on contrib/kafka/src/main/java/com/google/cloud/dataflow/contrib/kafka/KafkaIO.java:
```diff
@@ -0,0 +1,1046 @@
+
+      ConsumerRecords<byte[], byte[]> records;
+      try {
+        records = availableRecordsQueue.poll(10, TimeUnit.MILLISECONDS);
```
constant

dhalperi commented on contrib/kafka/src/main/java/com/google/cloud/dataflow/contrib/kafka/KafkaIO.java:
```diff
@@ -0,0 +1,1046 @@
+          if (consumed >= 0 && offset <= consumed) { // -- (a)
+            // this can happen when compression is enabled in Kafka (seems to be fixed in 0.10)
+            // should we check if the offset is way off from consumedOffset (say > 1M)?
+            LOG.warn("ignoring already consumed offset {} for {}", offset, pState.topicPartition);
```
this is fairly expected, right? Should we disable these warnings so that we don't log per element?

rangadi replied:
It is fixed in 10 and may be very near 0.9.x. It was essentially a bug in the consumer. It has enough information to skip these internally. 

I wish there was a way to expose such metrics through stats. I think we can leave it for now. One degenerate case (which I did hit) is when I reset kafka cluster while the dataflow job is running. Since the offset is reset back to 0, all the records were before the recently read offset

dhalperi commented on contrib/kafka/src/main/java/com/google/cloud/dataflow/contrib/kafka/KafkaIO.java:
```diff
@@ -0,0 +1,1046 @@
+    @Override
+    public boolean advance() throws IOException {
+      /* Read first record (if any). we need to loop here because :
+       *  - (a) some records initially need to be skipped if they are before consumedOffset
```
is it possible to move all the skipping logic into `start()`? Then `advance()` would be much cleaner I think.

rangadi replied:
We would have to loop in both cases. There is guarantee start() would be able to read all of them without blocking.

dhalperi commented on contrib/kafka/src/main/java/com/google/cloud/dataflow/contrib/kafka/KafkaIO.java:
```diff
@@ -0,0 +1,1046 @@
+          long offset = offsetConsumer.position(p.topicPartition);
+          p.setLatestOffset(offset);;
+        } catch (Exception e) {
+          LOG.warn(this + " : exception while fetching latest offsets. ignored.",  e);
```
don't construct strings with '+' in log statements, see other similar content. Please skim entire PR.

rangadi replied:
fixed.

dhalperi commented:
R: @dpmills @mshields822 

Daniel has already taken a look at this a month back, can you re-skim it?

Mark has been implementing PubSub. Mark, can you take a look at the Kafka source and see what lessons you can impart?

mshields822 commented on contrib/kafka/src/main/java/com/google/cloud/dataflow/contrib/kafka/KafkaIO.java:
```diff
@@ -0,0 +1,1046 @@
+import static com.google.common.base.Preconditions.checkNotNull;
+import static com.google.common.base.Preconditions.checkState;
+
+import com.google.api.client.repackaged.com.google.common.annotations.VisibleForTesting;
```
No need for repackaged.

rangadi replied:
oops! Thanks for catching.

mshields822 commented:
LGTM
No overlap with pub/sub impl:
 - Kafka: Due to strict partitioning cannot use additional source splits to naively hide get latency. So need background thread. Pubsub: No background threads needed.
 - Kafka: Since seekable (modulo caveats) can estimate backlog. Pubsub: Not  seekable and pub/sub don't publish any backlog estimate (at least publically). So backlog is just what's received but not yet read by advance in the reader.
 - Kafka: Imposes key/value structure on records. Pub/sub: uninterpreted bytes.
 - Kafka: Order preserving, so assuming (implied) timestamps are monotonic watermark is just current records timestamp. Pub/sub: Not order preserving, no assumption on timestamp monotonicity, need to estimate watermark.

rangadi commented:
Thanks @mshields822. 
One thing I would like to expand is that Kafka wouldn't be much different w.r.t timestamp and wantermarks. The order is only preserved within a partition, and we don't guarantee synchronous progress across multiple partitions.

dhalperi commented:
Looks like you need to push? I'm going to stop commenting on outstanding feedback :)

rangadi commented:
sorry, forgot to push before heading to lunch. Addressed all the comments except last one. I will double check any remaining comments.

dhalperi commented on contrib/kafka/src/main/java/com/google/cloud/dataflow/contrib/kafka/KafkaIO.java:
```diff
@@ -0,0 +1,1053 @@
+          }
+
+          ConsumerRecord<byte[], byte[]> rawRecord = pState.recordIter.next();
+          long consumed = pState.consumedOffset;
```
Is there a bound on the number of records that we'll have to skip? We can only log a few 1000 per second, so if it's millions then logging each record will induce large delays in starting up.

Alternative would be to set a flag and only log once.

rangadi replied:
It would fairly small (10s to 100s). The upper limit is the number of kafka records compressed together. Future versions of KafkaConsumer already fixed this issue with compressed messages and they skip them. So it is not at all expected in near future.

One case where it can cause millions of records is the case I mention above where Kafka is restarted from scratch which resets the offsets while the Dataflow app is running. Not sure weather we want to handle that case. It is better for the user to realize this problem (should be very rare), and take appropriate action (may be just restarting the Dataflow app).

That said I can certainly add a flag to limit it to one message.

dhalperi commented:
Closing in favor of https://github.com/apache/incubator-beam/issues/142
